File: starter_kits/Swift/Halite-III/Halite-III/hlt/Shipyard.swift
//
//  Shipyard.swift
//  Halite-III
//
//  Created by Chris Downie on 10/24/18.
//  Copyright © 2018 Chris Downie. All rights reserved.
//

import Foundation

struct Shipyard: Placeable, Structure {
    typealias ID = Int
    
    let type = StructureType.shipyard
    
    let owner: Player.ID
    let id: ID
    let position: Position
    
    /// Returns an engine command to generate a new ship.
    ///
    /// - Returns: An engine command to generate a new ship.
    func spawn() -> Command {
        return .generate
    }
}

File: starter_kits/Clojure/hlt/lib/hlt/shipyard.clj
(ns hlt.shipyard
  (:require [hlt.commands :as commands]))

(defn id
  "Get id for a shipyard"
  [shipyard]
  (::id shipyard))

(defn owner
  "Get owner for a shipyard"
  [shipyard]
  (:hlt.player/id shipyard))

(defn position
  "Get position of a shipyard"
  [shipyard]
  (:position shipyard))

(defn spawn
  "returns the command to spawn a ship"
  []
  commands/generate)

(defn new-shipyard
  "New shipyard"
  [owner shipyard-id position]
  {::id shipyard-id
   :hlt.player/id owner
   :position position})
